File: selfstudy/js/cars_avg.js
/*
task 24:
Cars - Avg
You need to compute the average of car speeds in km/h
*/

function average_cars(speeds) {
  return speeds.length ? speeds.reduce((a, b) => a + b, 0) / speeds.length : 0;
}

// Tests
const tests = [
  { input: [10, 20, 30], expected: average_cars([10, 20, 30]) },
  { input: [50, 60, 70], expected: average_cars([50, 60, 70]) },
  { input: [1, 2, 3], expected: average_cars([1, 2, 3]) },
  { input: [100, 200], expected: average_cars([100, 200]) },
  { input: [], expected: average_cars([]) }
];

if (process.argv.includes('mocha')) {
  const assert = require('assert');
  describe('average_cars', () => {
    tests.forEach((t, idx) => {
      it(`test case ${idx+1}`, () => {
        assert.deepStrictEqual(average_cars(t.input), t.expected);
      });
    });
  });
} else {
  tests.forEach((t, idx) => {
    console.log(`Test ${idx+1}:`, average_cars(t.input) === t.expected || JSON.stringify(average_cars(t.input)) === JSON.stringify(t.expected) ? "PASS" : "FAIL");
  });
}

// finally works

File: selfstudy/golang/cars_avg.go
/*
task 224:
Cars - Avg
You need to compute the average of car speeds in km/h
*/

package main

import (
	"fmt"
	"reflect"
)

func AverageCars(speeds []int) float64 {
	if len(speeds) == 0 { return 0 }
	sum := 0
	for _, v := range speeds {
		sum += v
	}
	return float64(sum)/float64(len(speeds))
}

func main() {
	tests := []struct {
		input []int
		expected float64
	}{
		{[]int{10, 20, 30}, AverageCars([]int{10, 20, 30})},
		{[]int{50, 60, 70}, AverageCars([]int{50, 60, 70})},
		{[]int{1, 2, 3}, AverageCars([]int{1, 2, 3})},
		{[]int{100, 200}, AverageCars([]int{100, 200})},
		{[]int{}, AverageCars([]int{})},
	}
	
	for idx, t := range tests {
		res := AverageCars(t.input)
		status := "FAIL"
		if reflect.DeepEqual(res, t.expected) {
			status = "PASS"
		}
		fmt.Printf("Test %d: %s\n", idx+1, status)
	}
}

// works i think
